src/object.rs

```diff
@@ -714,4 +717,56 @@ impl<T: IsA<Object> + SetValue> ObjectExt for T {
             }
         }
     }
+
+    fn downgrade(&self) -> WeakRef<T> {
+        unsafe {
+            let w = WeakRef(Box::new(mem::uninitialized()), PhantomData);
+            gobject_ffi::g_weak_ref_init(mut_override(&*w.0), self.to_glib_none().0);
+            w
+        }
+
+    }
+}
+
+pub struct WeakRef<T: IsA<Object> + ?Sized>(Box<gobject_ffi::GWeakRef>, PhantomData<*const T>);
+
+impl<T: IsA<Object> + Cast + ?Sized> WeakRef<T> {
+    pub fn upgrade(&self) -> Option<T> {
+        unsafe {
+            let ptr = gobject_ffi::g_weak_ref_get(mut_override(&*self.0));
+            if ptr.is_null() {
+                None
+            } else {
+                let obj: Object = from_glib_full(ptr);
+                Some(Cast::downcast::<T>(obj).expect("Wrongly typed GWeakRef"))
+            }
+        }
+    }
 }
+
+impl<T: IsA<Object> + ?Sized> Drop for WeakRef<T> {
+    fn drop(&mut self) {
+        unsafe {
+            gobject_ffi::g_weak_ref_clear(mut_override(&*self.0));
+        }
+    }
+}
+
+impl<T: IsA<Object> + ?Sized> Clone for WeakRef<T> {
+    fn clone(&self) -> Self {
+        unsafe {
+            let c = WeakRef(Box::new(mem::uninitialized()), PhantomData);
+
+            let o = gobject_ffi::g_weak_ref_get(mut_override(&*self.0));
+            gobject_ffi::g_weak_ref_init(mut_override(&*self.0), o);
+            if !o.is_null() {
+                gobject_ffi::g_object_unref(o);
+            }
+
+            c
+        }
+    }
+}
+
+unsafe impl<T: IsA<Object> + Send + Sync + ?Sized> Send for WeakRef<T> {}
+unsafe impl<T: IsA<Object> + Send + Sync + ?Sized> Sync for WeakRef<T> {}
```

Comment:
You need both to have each when reference counting is involved. We ensure that for Object anyway in gir, but better play safe here too.

For unsized, as a trait object maybe? Otherwise not really I guess. But doesn't hurt. I can remove that if you prefer that.